### PR TITLE
`make SPKG-check` for normal and script packages

### PR DESCRIPTION
**Pull request by mkoeppe**
`make SPKG-check` runs the testsuite of `SPKG` via sage-spkg. This builds on
- #36738
- #36747

### :memo: Checklist

- [ ] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: build/bin/sage-spkg [hunk 1]
```diff
@@ -584,14 +586,21 @@ if [ \$? -ne 0 ]; then
...
+    if [ -n "$fallback_script_dir" ]; then
+        cat >> "$tmpscript" <<__EOF__
+cd "\$SAGE_PKG_DIR" 2>/dev/null || cd "$fallback_script_dir"
+
+__EOF__
+    else
+        cat >> "$tmpscript" <<__EOF__
 if [ \$? -ne 0 ]; then
     echo >&2 "Error: could not cd to the package build directory \$SAGE_PKG_DIR"
     exit 1
```
Comment (kwankyu) [inline build/bin/sage-spkg:600]: Missing 
```
cd "\$SAGE_PKG_DIR"
```
before line 597 ?

File: build/make/Makefile.in [hunk 2]
```diff
@@ -760,22 +765,7 @@ $(1)-$(4)-no-deps:
...
-		PKG_BASE="$(1)" \
-		PKG_VER="$(2)" \
-		PKG_NAME="$(1)-$(2)" \
-		SAGE_SPKG_WHEELS=$$($(4))/var/lib/sage/wheels \
-		SAGE_SPKG_SCRIPTS=$$($(4))/var/lib/sage/scripts \
-		SAGE_INST_LOCAL=$$($(4)) \
-		SAGE_CHECK=$$(SAGE_CHECK_$(1)) \
-		sage-logger -p '$$(SAGE_ROOT)/build/pkgs/$(1)/spkg-check' '$$(SAGE_LOGS)/$(1)-$(2).log'; \
-	    fi
+	$(PLUS)@sage-logger -p 'PATH=$$(SAGE_SRC)/bin:$$($(4))/bin:$$$$PATH $$(SAGE_SPKG) --check-only $(1)-$(2) $$($(4))' '$$(SAGE_LOGS)/$(1)-$(2).log'
```
Comment (kwankyu) [inline build/make/Makefile.in:769]: Is this OK, removing all these lines?

File: build/bin/sage-spkg (hunk 1 above)
Comment (mkoeppe) [reply build/bin/sage-spkg:600]: no, it's in line 592

File: build/make/Makefile.in (hunk 2 above)
Comment (mkoeppe) [reply build/make/Makefile.in:769]: yes! that's a simplification that comes from using the `sage-spkg` program to run the package's scripts.

File: build/bin/sage-spkg (hunk 1 above)
Comment (kwankyu) [reply build/bin/sage-spkg:600]: Line 592 is in "if then" clause...

File: build/make/Makefile.in (hunk 2 above)
Comment (kwankyu) [reply build/make/Makefile.in:769]: I see.

File: build/bin/sage-spkg (hunk 1 above)
Comment (mkoeppe) [reply build/bin/sage-spkg:600]: oh, I have to look at this again with less tired eyes

Comment (kwankyu) [reply build/bin/sage-spkg:600]: Take your time.

Comment (mkoeppe) [reply build/bin/sage-spkg:600]: Of course, you were right.

**Comment by mkoeppe** [comment]
Ready to go?

**Comment by kwankyu** [comment]
I get
```
$ sage -i coxeter3
$ make coxeter3-check
make -j4 build/make/Makefile --stop
make[1]: `build/make/Makefile' is up to date.
build/bin/sage-logger \
		"cd build/make && ./install 'coxeter3-check'" logs/install.log
*** ALL ENVIRONMENT VARIABLES BEFORE BUILD: ***
BASH_COMPLETION_COMPAT_DIR=/usr/local/etc/bash_completion.d
BASH_SILENCE_DEPRECATION_WARNING=1
CLICOLOR=YES
...
__CFBundleIdentifier=com.apple.Terminal
__CF_USER_TEXT_ENCODING=0x1F5:0x0:0x0
***********************************************
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] Running the test suite for coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0...
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] /Users/kwankyu/GitHub/sage-dev/local/var/lib/sage/scripts/coxeter3/spkg-check: line 32: cd: src: No such file or directory
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] make[2]: *** No rule to make target `check'.  Stop.
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] 
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] real	0m0.420s
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] user	0m0.336s
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] sys	0m0.078s
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] ************************************************************************
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] Error testing package coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] ************************************************************************
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] Please email sage-devel (http://groups.google.com/group/sage-devel)
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] explaining the problem and including the log files
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0]   /Users/kwankyu/GitHub/sage-dev/logs/pkgs/coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0.log
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] and
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0]   /Users/kwankyu/GitHub/sage-dev/config.log
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] Describe your computer, operating system, etc.
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] If you want to try to fix the problem yourself, *don't* just cd to
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] /Users/kwankyu/GitHub/sage-dev/build/pkgs/coxeter3 and type 'make check' or whatever is appropriate.
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] Instead, the following commands setup all environment variables
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] correctly and load a subshell for you to debug the error:
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0]   (cd '/Users/kwankyu/GitHub/sage-dev/build/pkgs/coxeter3' && '/Users/kwankyu/GitHub/sage-dev/sage' --buildsh)
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] When you are done debugging, you can type "exit" to leave the subshell.
[coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0] ************************************************************************
make[1]: *** [coxeter3-SAGE_LOCAL-check] Error 1

real	0m0.721s
user	0m0.515s
sys	0m0.192s
***************************************************************
Error building Sage.

The following package(s) may have failed to build (not necessarily
during this run of 'make coxeter3-check'):

* package:         coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0
  last build time: Mar 22 08:47
  log file:        /Users/kwankyu/GitHub/sage-dev/logs/pkgs/coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0.log

It is safe to delete any log files and build directories, but they
contain information that is helpful for debugging build problems.
WARNING: If you now run 'make' again, the build directory of the
same version of the package will, by default, be deleted. Set the
environment variable SAGE_KEEP_BUILT_SPKGS=yes to prevent this.

make: *** [coxeter3-check] Error 1
```

**Comment by kwankyu** [comment]
I get
```
(after full make)
$ make cython-check
make -j4 build/make/Makefile --stop
make[1]: `build/make/Makefile' is up to date.
build/bin/sage-logger \
		"cd build/make && ./install 'cython-check'" logs/install.log
*** ALL ENVIRONMENT VARIABLES BEFORE BUILD: ***
BASH_COMPLETION_COMPAT_DIR=/usr/local/etc/bash_completion.d
BASH_SILENCE_DEPRECATION_WARNING=1
CLICOLOR=YES
CONDA_EXE=/usr/local/Caskroom/miniconda/base/bin/conda
CONDA_PYTHON_EXE=/usr/local/Caskroom/miniconda/base/bin/python
CONDA_SHLVL=0
...
__CFBundleIdentifier=com.apple.Terminal
__CF_USER_TEXT_ENCODING=0x1F5:0x0:0x0
***********************************************
[cython-3.0.7] The test suite for cython-3.0.7 cannot be run because the script
[cython-3.0.7] /Users/kwankyu/GitHub/sage-dev/local/var/lib/sage/venv-python3.12/var/lib/sage/scripts/cython/spkg-check
[cython-3.0.7] is missing. Install/re-install package cython-3.0.7 to run the test suite.
make[1]: *** [cython-SAGE_VENV-check] Error 1

real	0m0.371s
user	0m0.212s
sys	0m0.145s
***************************************************************
Error building Sage.

The following package(s) may have failed to build (not necessarily
during this run of 'make cython-check'):

* package:         coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0
  last build time: Mar 22 08:47
  log file:        /Users/kwankyu/GitHub/sage-dev/logs/pkgs/coxeter3-8ac9c71723c8ca57a836d6381aed125261e44e9e.p0.log

It is safe to delete any log files and build directories, but they
contain information that is helpful for debugging build problems.
WARNING: If you now run 'make' again, the build directory of the
same version of the package will, by default, be deleted. Set the
environment variable SAGE_KEEP_BUILT_SPKGS=yes to prevent this.

make: *** [cython-check] Error 1
```

**Comment by kwankyu** [comment]
OK. 
```
$ sage -f -s -c coxeter3
```
did it. I needed `-s` to keep the temporary build directory. 

I wonder why `-c` does not imply `-s` since `sage -f -c coxeter3` won't work.

Which spkg is a script package (with `spkg-check.in`)? It is hard to find one.

**Comment by mkoeppe** [comment]
> `sage -f -c coxeter3` won't work

What do you observe? For me it builds and tests just fine.

**Comment by mkoeppe** [comment]
The behavior for `cython` that you showed in https://github.com/sagemath/sage/pull/37022#issuecomment-2014063157 is as designed.

**Comment by mkoeppe** [comment]
> Which spkg is a script package (with `spkg-check.in`)?

On this branch there's no script package that has `spkg-check.in` but there are a few that have `spkg-check`:
```
$ ./sage -package list --has-file "spkg-check|spkg-check.in" --no-file checksums.ini 
sage_sws2rst
sagemath_categories
sagemath_environment
sagemath_objects
sagemath_repl
```

**Comment by kwankyu** [comment]
> > `sage -f -c coxeter3` won't work
> 
> What do you observe? For me it builds and tests just fine.

Yes. It works well. It tests first, and then deletes the temporary build directory.

What doesn't work is: `sage -f -c coxeter3 && make coxeter3-check`

**Comment by mkoeppe** [comment]
> What doesn't work is: `sage -f -c coxeter3 && make coxeter3-check`

Yes, that's as designed. However, if the test suite fails, then you can re-run it with `make coxeter3-check`

**Comment by kwankyu** [comment]
`sage -f sage_sws2rst && make sage_sws2rst-check` works since `$PKG_SCRIPTS` contains the `src` directory.

**Comment by kwankyu** [comment]
`make sagemath_categories` fails for me:
```
...
[sagemath_objects-10.3] ERROR: Could not find a version that satisfies the requirement sage-setup~=10.3 (from versions: none)
[sagemath_objects-10.3] ERROR: No matching distribution found for sage-setup~=10.3
...
```
So I cannot test this PR with this spkg.

Anyway this is not related with this PR.

**Review by kwankyu — APPROVED** [review]

**Comment by mkoeppe** [comment]
Thanks for the review!

**Comment by mkoeppe** [comment]
> `make sagemath_categories` fails for me

Yes, for that to work, one would need to do something like `make pypi-wheels` first

**Comment by kwankyu** [comment]
> > `make sagemath_categories` fails for me
> 
> Yes, for that to work, one would need to do something like `make pypi-wheels` first

That worked. Then from `make sagemath_categories-check
`, I got
```
[sagemath_categories-10.3] Running the test suite for sagemath_categories-10.3...
[sagemath_categories-10.3] Running tox -r -p auto -v --installpkg /Users/kwankyu/GitHub/sage-dev/local/var/lib/sage/venv-python3.12/var/lib/sage/wheels/sagemath_categories-10.3-cp312-cp312-macosx_12_0_x86_64.whl
[sagemath_categories-10.3]   sagepython-sagewheels-nopypi-norequirements: SKIP (1.03 seconds)
[sagemath_categories-10.3]   evaluation failed :( (1.15 seconds)
[sagemath_categories-10.3] New failures, not in baseline known-test-failures*.json
```
which is successful as a test :-)

**Comment by mkoeppe** [comment]
Great, thanks for testing!

**Pull request merged**